test/parallel/test-stream-readable-flow-no-flow-mixed.js

```diff
@@ -0,0 +1,102 @@
+'use strict';
+
+const common = require('../common');
+
+// This test ensures that Readable stream switches between flowing and
+// non-flowing state properly when varying the 'readable' and 'data' event
+// subscription.
+
+const assert = require('assert');
+const { Readable } = require('stream');
+
+const flowingData = [
+  { value: 'a' },
+  { value: 'b' },
+  { value: 'c', subscribeData: true },
+  { value: 'd' },
+  { value: 'e' },
+  { value: 'f', removeReadable: true },
+  { value: 'g' },
+  { value: 'h' },
+  { value: 'i', subscribeReadable: true },
+  null,
+];
+
+const r = new Readable({
+  read: common.mustCall(() => {
+    process.nextTick(() => {
+      r.push(flowingData.shift());
+    });
+  }, flowingData.length),
+  objectMode: true,
+
+  // The water mark shouldn't matter but we'll want to ensure the stream won't
+  // buffer data before we have a chance to react to the subscribe/unsubscribe
+  // event controls.
+  highWaterMark: 0,
+});
+
+// Store data received through 'readable' events and 'data' events.
+const actualReadable = [];
+const actualData = [];
+
+r.on('end', common.mustCall(() => {
+  assert.deepStrictEqual(actualReadable, ['a', 'b', 'c', 'd', 'e', 'f']);
+  assert.deepStrictEqual(actualData, ['d', 'e', 'f', 'g', 'h', 'i']);
+}));
+
+// Subscribing 'readable' should set flowing state to false.
+assert.strictEqual(r.readableFlowing, null);
+r.on('readable', common.mustCall(() => {
+  const v = r.read();
+  actualReadable.push(v.value);
+
+  if (v.subscribeData) {
+
+    // Subsribing 'data' should not change flowing state.
+    assert.strictEqual(r.readableFlowing, false);
+    r.on('data', common.mustCall((data) => {
```

Comment:
Can you move this `on('data')` listener to the top level. I would like to make sure that the stream is actually waiting for a call to `.read()` to emit `'data'`.